### PR TITLE
Fix several unused warnings (continued)

### PR DESCRIPTION
**Pull request by SuperV1234**
Resume work on #1791.

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by Bromeon** [comment]
Thanks a lot for continuing this!

It's currently a bit hard to easily see the additional changes, would it be possible for you to squash the commits into one (or a handful, if they are really logically separate)?

**Comment by SuperV1234** [comment]
> Thanks a lot for continuing this!
> 
> It's currently a bit hard to easily see the additional changes, would it be possible for you to squash the commits into one (or a handful, if they are really logically separate)?

Assuming you want to see only *my* changes, you can already do that:
https://github.com/SFML/SFML/pull/1846/files/77750f4fe89505b4c9b2a39c143dff490798f7eb..2e4b32d2899060af9f43a3c08c96df9685067431

Let me know if that works for you, otherwise I can squish.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by binary1248** [comment]
The following patch should solve the issues with Android/RPi.
```diff
diff --git a/src/SFML/Main/CMakeLists.txt b/src/SFML/Main/CMakeLists.txt
index 9344ada1..babd2b32 100644
--- a/src/SFML/Main/CMakeLists.txt
+++ b/src/SFML/Main/CMakeLists.txt
@@ -18,7 +18,7 @@ sfml_add_library(sfml-main STATIC SOURCES ${SRC})

 if(SFML_OS_ANDROID)
     # glad sources
-    target_include_directories(sfml-main PRIVATE "${PROJECT_SOURCE_DIR}/extlibs/headers/glad/include")
+    target_include_directories(sfml-main SYSTEM PRIVATE "${PROJECT_SOURCE_DIR}/extlibs/headers/glad/include")
 endif()

 # overwrite sfml-main suffix for backward compatibility with FindSFML.cmake
diff --git a/src/SFML/Network/IpAddress.cpp b/src/SFML/Network/IpAddress.cpp
index 0f0e9fb8..e7806407 100644
--- a/src/SFML/Network/IpAddress.cpp
+++ b/src/SFML/Network/IpAddress.cpp
@@ -198,7 +198,9 @@ void IpAddress::resolve(const std::string& address)
             {
                 if (result)
                 {
-                    ip = reinterpret_cast<sockaddr_in*>(result->ai_addr)->sin_addr.s_addr;
+                    sockaddr_in sin;
+                    std::memcpy(&sin, result->ai_addr, sizeof(result->ai_addr));
+                    ip = sin.sin_addr.s_addr;
                     freeaddrinfo(result);
                     m_address = ip;
                     m_valid = true;
```
Since the GL loaders are auto-generated, I wouldn't make changes to them unless absolutely necessary. Completely removing unused functions is easy but it gets tricky when modifying code inside function bodies and having to reapply the changes the next time the loaders need to be generated again.

**Comment by binary1248** [comment]
SFMLBuildMaster: Build this please.

**Comment by binary1248** [comment]
SFMLBuildMaster: Build this please.

**Comment by binary1248** [comment]
SFMLBuildMaster: Build this please.

**Comment by binary1248** [comment]
And this patch should fix the current alignment warning on RPi:
```diff
diff --git a/examples/voip/Server.cpp b/examples/voip/Server.cpp
index 0875bf95..bfeea8c0 100644
--- a/examples/voip/Server.cpp
+++ b/examples/voip/Server.cpp
@@ -126,14 +126,15 @@ private:
             if (id == audioData)
             {
                 // Extract audio samples from the packet, and append it to our samples buffer
-                const sf::Int16* samples     = reinterpret_cast<const sf::Int16*>(static_cast<const char*>(packet.getData()) + 1);
-                std::size_t      sampleCount = (packet.getDataSize() - 1) / sizeof(sf::Int16);
+                std::size_t sampleCount = (packet.getDataSize() - 1) / sizeof(sf::Int16);

                 // Don't forget that the other thread can access the sample array at any time
                 // (so we protect any operation on it with the mutex)
                 {
                     sf::Lock lock(m_mutex);
-                    std::copy(samples, samples + sampleCount, std::back_inserter(m_samples));
+                    std::size_t oldSize = m_samples.size();
+                    m_samples.resize(oldSize + sampleCount);
+                    std::memcpy(&(m_samples[oldSize]), static_cast<const char*>(packet.getData()) + 1, sampleCount * sizeof(sf::Int16));
                 }
             }
             else if (id == endOfStream)
```

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by binary1248** [comment]
I think the last conversion warnings in the Vulkan example are trivial enough to fix without my help. 😄

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: extlibs/headers/glad/include/glad/egl.h [hunk 1]
```diff
@@ -1344,12 +1328,6 @@ static int gladLoadEGLUserPtr(EGLDisplay display, GLADuserptrloadfunc load, void
     return version;
 }
 
-static int gladLoadEGL(EGLDisplay display, GLADloadfunc load) {
-    return gladLoadEGLUserPtr(display, sf_glad_egl_get_proc_from_userptr, GLAD_GNUC_EXTENSION (void*) load);
-}
-
-
-
```
Comment (eXpl0it3r) [inline extlibs/headers/glad/include/glad/egl.h:1352]: Were these changes required, @SuperV1234 and can/should we keep them, @binary1248?

File: extlibs/headers/glad/include/glad/wgl.h [hunk 2]
```diff
@@ -571,16 +567,6 @@ static int gladLoadWGLUserPtr(HDC hdc, GLADuserptrloadfunc load, void *userptr)
...
- 
-
-#ifdef SF_GLAD_WGL
-
-static int gladLoaderLoadWGL(HDC hdc) {
-    return gladLoadWGLUserPtr(hdc, sf_glad_wgl_get_proc_from_userptr, GLAD_GNUC_EXTENSION (void*) wglGetProcAddress);
-}
-
-
-#endif /* SF_GLAD_WGL */
```
Comment (eXpl0it3r) [inline extlibs/headers/glad/include/glad/wgl.h:583]: Were these changes required, @SuperV1234 and can/should we keep them, @binary1248?

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: extlibs/headers/glad/include/glad/egl.h (hunk 1 above)
Comment (binary1248) [reply extlibs/headers/glad/include/glad/egl.h:1352]: I removed the functions in my first commit. They are not used and would cause warnings on some compilers.

File: extlibs/headers/glad/include/glad/wgl.h (hunk 2 above)
Comment (binary1248) [reply extlibs/headers/glad/include/glad/wgl.h:583]: I removed the functions in my first commit. They are not used and would cause warnings on some compilers.

**Comment by eXpl0it3r** [comment]
Anyone wanna review all the changes? 😄

**Review by binary1248 — APPROVED** [review]

**Comment by binary1248** [comment]
Looks good to me. 👍

File: src/SFML/Window/GlContext.cpp [hunk 3]
```diff
@@ -291,14 +291,14 @@ namespace
...
-                major = version[prefixLength] - '0';
-                minor = version[prefixLength + 2] - '0';
+        if ((std::strlen(version) >= (prefixLength + 3)) &&
+            (std::strncmp(version, prefix, prefixLength) == 0) &&
+            std::isdigit(version[prefixLength]) &&
+            (version[prefixLength + 1] == '.') &&
+            std::isdigit(version[prefixLength + 2]))
+        {
+            major = static_cast<unsigned int>(version[prefixLength] - '0');
+            minor = static_cast<unsigned int>(version[prefixLength + 2] - '0');
```
Comment (kimci86) [inline src/SFML/Window/GlContext.cpp:301]: One indentation level is missing.

File: src/SFML/Window/Unix/WindowImplX11.cpp [hunk 4]
```diff
@@ -81,10 +82,10 @@ namespace
...
         static const unsigned int             maxTrialsCount = 5;
 
-        // Predicate we use to find key repeat events in processEvent
-        struct KeyRepeatFinder
-        {
-            KeyRepeatFinder(unsigned int keycode, Time time) : keycode(keycode), time(time) {}
+    // Predicate we use to find key repeat events in processEvent
+    struct KeyRepeatFinder
+    {
+        KeyRepeatFinder(unsigned int initalKeycode, Time initialTime) : keycode(initalKeycode), time(initialTime) {}
```
Comment (kimci86) [inline src/SFML/Window/Unix/WindowImplX11.cpp:88]: One indentation level is missing.

File: src/SFML/Window/Unix/WindowImplX11.cpp [hunk 5]
```diff
@@ -117,11 +118,11 @@ namespace
...
-            while ((result = read(file, &buffer[offset], 256)) > 0)
-            {
-                buffer.resize(buffer.size() + result, 0);
-                offset += result;
-            }
+        while ((result = read(file, &buffer[offset], 256)) > 0)
+        {
+            buffer.resize(buffer.size() + static_cast<std::size_t>(result), 0);
+            offset += static_cast<std::size_t>(result);
+        }
```
Comment (kimci86) [inline src/SFML/Window/Unix/WindowImplX11.cpp:125]: Also here, one indentation level is missing.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
Thanks a lot to everyone who spent their time on this sometimes rather tedious work! 🙂 
Having warnings treated as errors will help us a lot to keep things explicit and prevent potential errors.
While many warnings were now mostly silenced with explicit conversions/casts, there is probably still some more room to improve, so future code changes in an area, can/should also consider whether there's a better way to implement it without casts.

**Comment by Bromeon** [comment]
Thanks to everyone as well, and especially you @eXpl0it3r for managing the whole process!
Glad to see this merged, will definitely avoid quite some disruptive merge conflicts 🙂 

On the downside, code got quite a bit less readable through the ubiquitious casts. It also doesn't help that C++'s `static_cast` syntax is very ugly. And with all the different system libraries we have to interact, each with their own convention, it's hardly possible to avoid conversions, so it's probably still better to have them explicit.

**Comment by texus** [comment]
My CI has picked up a crash in `sf::Texture::copyToImage` that seems to have been introduced in [01836cc](https://github.com/SFML/SFML/commit/01836ccea4f5655e16ad6d7cf4aac6f023234c53#diff-d0c8f10fc212191720d2b7629b54b7e0bbeaa264759b1a386a5a0b7a5f15a559R374).

It looks like srcPitch needs to be a negative number if m_pixelsFlipped is true. The change to unsigned int causes a segmentation fault in memcpy.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,5 +1,8 @@
 cmake_minimum_required(VERSION 3.7.2)
 
+# customize the compiler options CMake uses to initialize variables with
+set(CMAKE_USER_MAKE_RULES_OVERRIDE "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CompilerOptionsOverride.cmake")
+
```
Comment (ChrisThrasher) [inline CMakeLists.txt:5]: What's the purpose of this variable and the corresponded file added? I read through Kitware's [documentation](https://cmake.org/cmake/help/v3.8/variable/CMAKE_USER_MAKE_RULES_OVERRIDE.html) of it and still can't understand how SFML is benefiting from it. For what it's worth CI still passes even if you remove this.

Comment (ChrisThrasher) [reply CMakeLists.txt:5]: @binary1248 I see you added this in a759ccd5d. Can you help us understand what purpose it serves?

Comment (binary1248) [reply CMakeLists.txt:5]: Without it MSVC would complain about duplicate warning flags because a warning flag is already part of the initial set of flags set by CMake.
